proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+
+  public subscript(_ position: Int) -> dependsOn(self) Element { get }
+}
+```
+
+##### Accessing subranges of elements:
+
+In SE-0437, `UnsafeBufferPointer`'s slicing subscript was replaced by the `extracting(_ bounds:)` functions, due to the copyability assumption baked into the standard library's `Slice` type. `Span` has similar requirements as `UnsafeBufferPointer`, in that it must be possible to obtain an instance representing a subrange of the same memory as another `Span`. We therefore follow in the footsteps of SE-0437, and add a family of `extracting()` functions:
+
+```swift
+extension Span {
+  func extracting(_ bounds: Range<Int>) -> dependsOn(self) Self
+  func extracting(_ bounds: some RangeExpression<Int>) -> dependsOn(self) Self
+  func extracting(_: UnboundedRange) -> dependsOn(self) Self
+}
+```
+
+##### Unchecked access to elements and subranges of elements:
+
+The `subscript` and the `extracting()` functions mentioned above all have always-on bounds checking of their parameters, in order to prevent out-of-bounds accesses. We also want to provide unchecked variants as an alternative for cases where bounds-checking is proving costly.
+
+```swift
+extension Span {
+  // Unchecked subscripting and extraction
+
+  /// Accesses the element at the specified `position`.
+  ///
+  /// This subscript does not validate `position`; this is an unsafe operation.
+  ///
+  /// - Parameter position: The position of the element to access. `position`
+  ///     must be a valid index that is not equal to the `endIndex` property.
+  public subscript(
+    unchecked position: Index
+  ) -> dependsOn(self) Element { get }
+
+  /// Accesses a contiguous subrange of the elements represented by this `Span`
+  ///
+  /// This subscript does not validate `bounds`; this is an unsafe operation.
+  ///
+  /// - Parameter bounds: A range of the collection's indices. The bounds of
+  ///     the range must be valid indices of the collection.
+  public extracting(
+    uncheckedBounds bounds: Range<Index>
+  ) -> dependsOn(self) Self
+
+  /// Accesses the contiguous subrange of the elements represented by
+  /// this `Span`, specified by a range expression.
+  ///
+  /// This subscript does not validate `bounds`; this is an unsafe operation.
+  ///
+  /// - Parameter bounds: A range of the collection's indices. The bounds of
+  ///     the range must be valid indices of the collection.
+  public subscript(
+    uncheckedBounds bounds: some RangeExpression<Index>
+  ) -> dependsOn(self) Span<Element>
+}
+```
+
+##### Index validation utilities:
+
+Every time `Span` uses a position parameter, it checks for its validity, unless the parameter is marked with the word "unchecked". The validation is performed with these functions:
+
+```swift
+extension Span {
+  /// Traps if `position` is not a valid offset into this `Span`
+  ///
+  /// - Parameters:
+  ///   - position: an position to validate
+  public boundsCheckPrecondition(_ position: Index)
```

Comment:
These aren't good names. Do we really want these to trap rather than returning a Bool?

Reply:
We want the trapping one internally, and I don't see a reason not to expose them, since a wrapper might be interested in using them. We should have the `Bool`-returning validating ones too.

Reply:
Added a `Bool`-returning bounds-checking function in addition to the (renamed) trapping one.